### DISCUSSION TITLE
Word processor

### DISCUSSION THREAD

**Discussion by ccoffing** [jbruchon/elks#1576] [General]
I have started work on a word processor for ELKS.  I'm writing in C99 and keeping it portable, and first and foremost keeping things 16-bit friendly.  I am doing my main development on big-brother-Linux and FreeBSD.

There are many many editors out there, but few finished and polished, yet even a good editor is not a word processor.  That's actually where this started, when I was trying to fix bugs in an editor on ELKS, and just got disgusted with code quality.  Why am I fixing this, when elvis is better, yet elvis isn't enough.

I have fond memories of Word Perfect 5.1, but have long since forgotten the function key mappings, but that's still pretty much the feel that I have in mind.  I want reveal codes.  Well, and I want the improved reflow from WP6.  I'm currently a daily vim user, but no matter how great of a tool it is, vim is not a word processor.  LibreOffice is great for what it is, but it's not a command-line tool.  I desire a terminal-friendly, fast-loading, word processor that can, at least a little, interface to the modern world (and ELKS).

I'm only a week in on this project, so energy is high, but of course will enter the pit of reality soon.  Bear with me.

Some of the features I'm hoping to (and have ideas how to) fit into 64K of code:
- Soft line wrapping and dynamic reflow
- Support "long" lines (not yet clear what that means, but not hardcoded and not a single allocation)
- Support "large" files (dynamic loading seems possible, but would require reliable storage for the temporary)
- Basic formatting (bold, italic, bullets, lists, etc)
- BMP unicode support, at least in principle, ignoring terminal limitations
- ELKS tty
- Maybe build-time selection between Emacs, vi, WordPerfect inspired key bindings

Some other (future!) things might be better done in old-school fashion with pipe and processes, such as import and export of formats.  Maybe spell check.  It would also be cool to keep the terminal details separate from the core, so that a nano-X build would be possible.  But these are just blue-sky ideas at this point, I could be full of shit.

Which then leads to other thoughts...

- I would like to keep this generic (a tool on the daily-driver) so a separate repo.  Probably then a manual periodic PR into ELKS?  I have wondered before what a ports tree would look like for ELKS.  That seems more scalable than stuffing everything in `elkscmd`.
- I hate licensing.
- The core doesn't seem deadly-hard, so why does everyone write small half-finished editors and not small word processors (I'll know once I step on the landmine)
- A small program like this seems like it would benefit from link-time optimization (which is basically like an auto-inline on steroids, right?) but I notice that the ELKS gcc does not support `-flto`.  I don't yet know if I care, but is there a technical reason this is not supported?

Too much talking.  Back to my cave.

**Comment by ghaerr** [comment]
Hello @ccoffing,

Well I can see you're quite excited about writing a word processor! I would think that if you could get a simple core formatter written, the rest should follow in somewhat straightforward manner.

> I would like to keep this generic (a tool on the daily-driver) so a separate repo.

See my [FM File Manager](https://github.com/ghaerr/fm) project. I developed it on macOS and kept a seperate [Makefile.elks](https://github.com/ghaerr/fm/blob/master/Makefile.elks) for building an ELKS version. Eventually I copied the ELKS version over to elkscmd/fm, for ease of distribution.

> I have wondered before what a ports tree would look like for ELKS. That seems more scalable than stuffing everything in elkscmd.

For larger projects, I've found it easier to build outside ELKS, and have a version that runs on the host OS as well. As you can see in the fm/Makefile.elks makefile, I just created a few important rules and set them in makefile, rather than trying to connect in to the elkscmd/ build system, at least in the beginning.

> Probably then a manual periodic PR into ELKS?

You're welcome to send periodic PRs to ELKS as well (perhaps elkscmd/wp)?

> A small program like this seems like it would benefit from link-time optimization 

I'm not really familiar with LTO nor `-flto`. I'll have to look further into that.

Some comments on features:

> BMP unicode support, at least in principle, ignoring terminal limitations

See my work in elkscmd/tui, where `fm`, `matrix`, `cons` and a few other programs are built. It puts together a framework where 16-bit (thus BMP) Unicode is fully supported within the application itself, but then uses a translation layer during output to the terminal. Specifically, translations to CP 437 occur when executing on the ELKS console, but when outputting to the serial port (expecting a modern terminal emulator), Unicode is translated into UTF-8 sequences. A similar thing happens for color attributes.

Try running `matrix` to see how console vs serial display is handled. Obviously, if any special characters used in the application also have a code point in CP 437, things work out better on real PCs and the ELKS console. (Note: neither ELKS kernel nor the console driver currently actually support unicode or UTF-8. This could be added, but console output is limited to CP 437 regardless).

> ELKS tty

See above, ELKS and UNIX tty handled in elkscmd/tui.

> Basic formatting (bold, italic, bullets, lists, etc)

Use ANSI sequences for bold/italic etc (we can add to console if not yet implemented), and use CP 437 code points for bullets or lists if possible, at least in the beginning. This will allow operation on console and external terminals (and host OS).

Thank you!
